### PR TITLE
feat(shed-core): mobile-parity gap-fill — overview, sessions, rc-events, token FSM (Phase A)

### PR DESCRIPTION
**Pull request by charliek**
## What & why

Additive, **library-only** gap-fill of `shed-core` / `shed-app` — **Phase A** of porting the `shed-mobile` Flutter app onto the shared Rust client core (killing ~2,000 lines of hand-maintained Dart that already carry "Mirrors the Rust core" comments). This PR adds the client surface mobile actually uses so a later flutter_rust_bridge layer can consume it; every addition follows `crates/` conventions and is ported field-for-field from the Dart sources, so Swift (UniFFI) and Tauri can adopt the same code next.

**No release, no manifest bump, no `shed-core-ffi` API change, no new dependencies.** Desktop behavior is unchanged except three strictly-safer improvements (below). This is a **stacked-commit** PR — each commit is independently shippable and separately reviewable.

## Commits (per workstream)

| Commit | Adds |
|---|---|
| **C1** overview DTOs + RC groundwork | `Overview`/`OverviewServer`/`OverviewShed` with Dart-tolerant per-field decode (every substructure degrades independently); `RcActivity` (Unknown catch-all); activity fields on `RcSession`/`RcSessionDto` (`skip_serializing_if`); `RcKindFeatures.watch`/`input` (drift fix); `ServerInfo.features` + `has_feature()` (Go emits it; Rust was dropping it); `fixtures/overview.json` byte-copied from mobile's golden |
| **C2** sessions read-plane + rc proxy | `overview()`, `list_sessions()`, `delete_session()`, `rc_messages()`, `rc_input()`; `Session`/`SessionRC`/`RcMessagesPage` DTOs; status-code error dispatch |
| **C3** rc-events decode + fold | `rc_events` module: `RcEvent` tolerant never-error decode + `ActivityOverlay` fold (blocking-state suppression, per-shed drops), ported from `rc_events.dart` with its full test suite |
| **C4** rc-events SSE client | `Client::rc_events(sink)` — one connection, capped parser, **byte-level idle timer** (heartbeat comments reset it) |
| **C5** `RcEventsWatcher` (shed-app) | shared reconnect loop + fold + channel; `gotData` resync parity; backoff 500ms→30s; spawn-once/non-restartable |
| **C6** token FSM knobs + stale-401 guard | `with_now`/`with_refresh_window`/`with_name_jitter`/`with_mint_cooldown`/`with_seed` builders; keep-valid-on-proactive-failure; `invalidate_if_current()` |
| **C7** token-bundle fail-closed parse | `parse_token_bundle`/`parse_control_bundle` (mobile parity) + std-only RFC3339 parser (no chrono) |
| **C8** rc permission modes | permission-mode sets + `validate_permission_mode`; `--permission-mode` threaded through `create_argv`/`create_invocation` |

## Verification

**Automated (full workspace green):** shed-core **270** tests (+116 over branch base), shed-app 89 default / 102 `rc` / 121 `broker,rc`, plus shed-broker 339 / host-agent 43 / ffi 2 / shedctl 13 — 0 failures. Workspace + feature-matrix clippy `-D warnings` clean. Tauri `cargo check` green after C1's `RcSession` shape change. **Reverse-dep AC:** `cargo tree -i saphyr-parser|notify|aws-sdk-sts` show no path into shed-core / shed-core-ffi / default shed-app. No `Cargo.lock` change.

**Per-commit adversarial review:** every commit ran implement → gate → simplify → **Codex correctness review** → fix → commit. Real defects caught and fixed with regression tests each round — notably a **watcher receiver-drop leak** (C5: watcher + TCP connection lived forever on a heartbeat-only stream) and **5 auth findings** in C6 (unauthenticated retry on re-mint failure, `last_error` secret leak, same-token re-cache, unbounded invalidate await), plus a **path-segment injection** in C2 (`delete_session("proj","../../victim")` → `DELETE /api/sheds/victim`).

**Live wire-shape check** against the real localmac server (v0.7.6, *older* than the golden's v0.8.0 — a good backward-compat test): `/api/info` returned no `features` field (validates `ServerInfo.features` null_default); a live SSH `_bootstrap control` mint returned a bundle matching `parse_control_bundle`'s shape; `/api/overview` returned a `text/plain` `404 page not found` — the exact chi-mislabeled-404 `overview()` short-circuits to `BadStatus(404)` before decode. The rc-events/sessions **success** shapes don't exist on v0.7.6, so they're covered by the byte-copied golden + Go-handler citations + raw-TCP SSE streaming tests (honest coverage boundary noted below).

## Desktop-visible changes (all strictly safer)
1. Keep the cached control token on a proactive-mint failure while still valid (C6).
2. `invalidate_if_current(observed)` stale-401 guard — a 401 racing a rotation can't clear a newer token (C6).
3. All Client URL building one-segment-encodes identifiers + rejects bare `""`/`"."`/`".."` — closes a path-traversal class; byte-identical requests for all valid names (C2).

## Accepted risk (dispositioned)
Token-provider single-flight holds its mutex across the mint, so cancelling a `token()` waiter can start a rare second concurrent mint (one extra round-trip, one abandoned short-TTL token; never a lockout or wrong cached token). Documented on the type with shed-broker's watch-based single-flight as the future upgrade.

## Follow-ups (not this PR)
Tauri/Swift adopt `RcEventsWatcher`; FFI export of the new surface; shed-lifecycle push events; live re-verify against a v0.8.0+ server. Phase B (the shed-mobile FRB bridge) pins to this merge.

<details>
<summary>Plan (durable record — the plan lives outside the repo)</summary>

Full plan + §Verified: `~/.claude/plans/shed/001-shed-core-mobile-gaps.md` (panel-reviewed by Codex + GLM 5.2 + CodeRabbit before implementation; corrections recorded in its header). Parent program: `~/.claude/plans/shed-mobile-rust-core.md`.

Design decisions pinned in the plan: full-transport-swap adoption depth; rc-events split as pure-fold (shed-core) + reconnect-watcher (shed-app) so reconnect logic lives once; byte-level idle timer (heartbeat comments are SSE comments the parser swallows); Dart-tolerant overview decode across every substructure (forward-compat: a new server enum value must not vanish a session); identity-binding deliberately not ported (a new `Client` per transport identity deletes the race class); UTF-16 `name_jitter` for Dart parity; RFC3339 parser stricter-than-Dart (fail-closed) but leap-second-tolerant.
</details>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a richer server overview payload (server info, feature flags, storage totals, sheds, sessions, capabilities, warnings).
  * Added live RC event watching with typed updates, reconnection/backoff, and resynchronization behavior.
  * Added RC activity metadata, message/session feed decoding, and permission-mode support.

* **Bug Fixes**
  * Improved auth recovery with safer, bounded retries and hardened token handling.
  * Strengthened URL construction and SSE stream robustness (liveness, exact success handling, tolerant decoding).
  * Improved defensive parsing for overview/sessions and portability for peer ID detection.

* **Tests**
  * Expanded coverage for auth edge cases, SSE timing/frames, rc-events overlay/resync, and parsing tolerance.

### CODE REVIEW

File: crates/shed-core/src/rc_events.rs [hunk 1]
```diff
@@ -0,0 +1,822 @@
...
+            RcEvent::HubUnavailable { shed } | RcEvent::ShedStopped { shed } => {
+                self.drop_shed(shed)
+            }
+        }
+    }
+
+    fn with(&self, key: (String, String), v: LiveActivity) -> ActivityOverlay {
+        let mut patches = self.patches.clone();
+        patches.insert(key, v);
+        ActivityOverlay { patches }
```
[the remark this answers is not in the document]
Comment (charliek) [reply crates/shed-core/src/rc_events.rs:301]: Deliberately dispositioned as accepted risk (not fixing in this PR), for three reasons:

1. **Matches repo precedent.** The `mpsc::UnboundedReceiver` + stop-on-send-failure shape mirrors `shed-app/src/host_agent.rs`'s existing long-lived event client. Introducing a bounded/drop-policy channel here would diverge from the established pattern for one of several similar producers.

2. **Bounded in practice.** The overlay is keyed by `(shed, slug)`; keys only exist for live `rc-*` tmux sessions on running sheds, and the synthetic `hub.unavailable` / `shed.stopped` events drop a shed's entire key-set. A "unique-slug flood" would require the server to actually mint that many real RC sessions.

3. **The realistic consumer drains eagerly.** The Phase B flutter_rust_bridge consumer (and the future Tauri/Swift adopters) pull each `RcWatcherUpdate` straight through to UI state; a stalled consumer that stops draining is itself a bug that the `updates.closed()` teardown arm (added after the earlier adversarial review) already converts into watcher shutdown when the receiver is dropped.

This is recorded on the `ControlTokenProvider`-adjacent watcher docs and in the plan's risk section. The proper upgrade if it ever matters — a bounded channel with an explicit lag/drop policy plus overlay-size capping — is noted as future work.
